### DISCUSSION TITLE
Linker error with ccache 4.3 and arm-none-eabi-gcc

### DISCUSSION THREAD

**Discussion by dueringa** [ccache/ccache#829] [Support]
Hi,

I'm developing firmware for an NRF52832 using

```
arm-none-eabi-gcc.exe (GNU Tools for Arm Embedded Processors 7-2018-q2-update) 7.3.1 20180622 (release) [ARM/embedded-7-branch revision 261907]
```

The firmware consists of several build configurations (with minor difference in the defines/compiler flags). The files are shared across the configurations.

I'm using ccache 4.2 under Windows (downloaded from the homepage) and encountered several different linker errors as soon as I start building the second configuration. I tried clearing the cache, but the linker error happened again.

The linker errors I encountered were both of the sort "undefined reference to..." as well as "multiple definition of...". The same function name is used multiple times in several compilation units, but the functions actually should be excluded by the preprocessor. So:

```c
// first.c
#if SYSTEM_ONE
int main()
{
//...
}
#endif

// second.c
#if SYSTEM_TWO
int main()
{
//...
}
#endif
```

Compilation works fine without CCACHE.

The command line tends to become very long (3600 characters) due to lots of include directories and defines. Maybe I'm hitting a limit here?

The scenatio happens both with parallel make (-j4) as well as single process make (-j1).

However, using ccache 3.7.7, everything works fine. (Using the build from http://repo.msys2.org/mingw/x86_64/mingw-w64-x86_64-ccache-3.7.7-1-any.pkg.tar.xz)

I could test the other versions between those as well, before I do that, any ideas what could cause these linker errors, and how I could debug this?

(Unfortunately, I can't share the code causing these issues, since it's proprietary).

**Comment by jrosdahl** [comment]
Sounds like it might be due to the bug that #816 fixed. It will be included in version 4.2.1.

**Reply by dueringa** [reply]
Thanks a lot. This version fixed the issues indeed :+1:
